### PR TITLE
Fix issue #9202 - mcrz

### PR DESCRIPTION
**Pull request by adjs**
### Summary

This PR is on top of PR #9688 and fix issue #9202.

### CODE REVIEW

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py [hunk 1]
```diff
@@ -197,16 +305,14 @@ def mcry(
...
     else:
         raise QiskitError(f"Unrecognized mode for building MCRY circuit: {mode}.")
 
 
 def mcrz(
-    self,
-    lam: ParameterValueType,
-    q_controls: Union[QuantumRegister, List[Qubit]],
-    q_target: Qubit,
-    use_basis_gates: bool = False,
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py:209]: We'll have to deprecate this behavior before removing it. If the new code uses different gates, one option could be to call `transpile` with `optimization_level=0` and `basis_gates=["p", "u", "cx"]` on the decomposition.

To deprecate this argument, you could check out this example: https://github.com/Qiskit/qiskit-terra/blob/a461a913c2e5bd51b307f38b89d2ab37117b7bf4/qiskit/algorithms/amplitude_amplifiers/grover.py#L117

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py [hunk 2]
```diff
@@ -230,13 +337,11 @@ def mcrz(
     self._check_dups(all_qubits)
 
     n_c = len(control_qubits)
-    if n_c == 1:  # cu
-        _apply_cu(self, 0, 0, lam, control_qubits[0], target_qubit, use_basis_gates=use_basis_gates)
+    if n_c == 1:
+        self.compose(CRZGate(lam), control_qubits + [target_qubit], inplace=True)
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py:341]: I think `append` would be the more adequate function to use here, as it's specifically for instructions/gates 🙂 
```suggestion
        self.append(CRZGate(lam), control_qubits + [target_qubit])
```

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py [hunk 3]
```diff
@@ -230,13 +337,11 @@ def mcrz(
...
     else:
-        lam_step = lam * (1 / (2 ** (n_c - 1)))
-        _apply_mcu_graycode(
-            self, 0, 0, lam_step, control_qubits, target_qubit, use_basis_gates=use_basis_gates
-        )
+        # self.append(u1_gate.control(n_c), control_qubits + [target_qubit])
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py:343]:
```suggestion
```

File: releasenotes/notes/su2-synthesis-295ebf03d9033263.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---
+features:
+  - |
+    The new function `mcsu2_real_diagonal(circuit, unitary, controls, target, ctrl_state)` 
+    applies a multi-controlled SU(2) gate unitary with one real diagonal. With n-1 controls, 
+    the decomposition requires approximately 16n cx gates. Synthesis of mcrx and mcry without 
+    auxiliary qubits now uses the `mcsu2_real_diagonal` decomposition.
```
Comment (Cryoris) [inline releasenotes/notes/su2-synthesis-295ebf03d9033263.yaml:7]: Since this function is not in the API documentation, I think we don't have to mention it in the releasenotes. Instead we could rephrase this as bugfix. Also, we should include the deprecation. Here's an example, but feel free to rephrase as is best!
```suggestion
fixes:
  - |
    Fixed the gate decomposition of multi-controlled Z rotation gates added via
    :meth:`.QuantumCircuit.mcrz`. Previously, this method implemented a multi-controlled
    phase gate, which has a relative phase difference  to the Z rotation.
deprecation:
  - |
    Deprecated the `use_basis_gates` argument in the multi-controlled Pauli rotation methods,
    :meth:`.QuantumCircuit.mcrx`, :meth:`.QuantumCircuit.mcry` and :meth:`.QuantumCircuit.mcrz`.
    To convert to a certain basis, you can instead use the :func:`.transpile` function with the 
    ``basis_gates`` argument, or the :class:`.BasisTranslator` transpiler pass.
```

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py [hunk 5]
```diff
@@ -81,6 +83,106 @@ def _apply_mcu_graycode(circuit, theta, phi, lam, ctls, tgt, use_basis_gates):
         last_pattern = pattern
 
 
+def mcsu2_real_diagonal(
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py:86]: Maybe it would be better to mark this function as private as it is a utility and not API documented. If we want to expose it, it might be worth to turn it into it's own class in a follow-up PR.

**Review by Cryoris — COMMENTED** [review]
Thanks for this addition, overall the fix looks great! I left some comments below, mainly formalities. 

If possible I think we would like to include this in the upcoming release, so let us know if there's anything we can help you with 🙂

**Comment by ShellyGarion** [comment]
- Why don't the gates `mcrx`, `mcry` and `mcrz` appear in the circuit library? 
https://qiskit.org/documentation/apidoc/circuit_library.html
Should they be added to the init file? 
https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/circuit/library/__init__.py

- Why do the `mcry` gate have a `q_ancillae` parameter and `mcrx` and `mcrz` do not have it?

File: qiskit/circuit/add_control.py [hunk 6]
```diff
@@ -146,8 +146,8 @@ def control(
                     gate.definition.data[0].operation.params[0],
                     q_control,
                     q_target[bit_indices[qargs[0]]],
-                    use_basis_gates=True,
                 )
+                continue
```
[the remark this answers is not in the document]
Comment (adjs) [reply qiskit/circuit/add_control.py:151]: Yes, because the decomposition is phase correct and the fragment below will change the global phase.

```python
            if gate.definition is not None and gate.definition.global_phase:
                global_phase += gate.definition.global_phase
```

File: qiskit/circuit/add_control.py [hunk 7]
```diff
@@ -184,23 +184,17 @@ def control(
                             use_basis_gates=True,
                         )
                     elif theta == 0 and phi == 0:
-                        controlled_circ.mcrz(
-                            lamb, q_control, q_target[bit_indices[qargs[0]]], use_basis_gates=True
-                        )
+                        controlled_circ.mcp(lamb, q_control, q_target[bit_indices[qargs[0]]])
```
[the remark this answers is not in the document]
Comment (adjs) [reply qiskit/circuit/add_control.py:188]: It still has no auxiliary qubits, however the basis_gates are different. This is an internal change and probably doesn't need to be reported in the release notes.

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py [hunk 8]
```diff
@@ -81,6 +83,106 @@ def _apply_mcu_graycode(circuit, theta, phi, lam, ctls, tgt, use_basis_gates):
...
+    Raises:
+        QiskitError: parameter errors
+    """
+    # pylint: disable=cyclic-import
+    from qiskit.circuit.library import MCXVChain
+    from qiskit.extensions import UnitaryGate
+    from qiskit.quantum_info.operators.predicates import is_unitary_matrix
+
+    if not is_unitary_matrix(unitary):
+        raise QiskitError("parameter unitary in linear_depth_mcv must be an unitary matrix")
```
[the remark this answers is not in the document]
Comment (adjs) [reply qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py]: This PR is on top of PR https://github.com/Qiskit/qiskit-terra/pull/9688. PR #9688 was merged and this modification was performed on it.

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py [hunk 9]
```diff
@@ -101,6 +203,8 @@ def mcrx(
     Raises:
         QiskitError: parameter errors
     """
+    from .rx import RXGate
```
[the remark this answers is not in the document]
Comment (adjs) [reply qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py]: Must be local imports to avoid cyclic imports errors.

File: releasenotes/notes/su2-synthesis-295ebf03d9033263.yaml (hunk 4 above)
Comment (adjs) [reply releasenotes/notes/su2-synthesis-295ebf03d9033263.yaml]: This release note appears in #9688, I have written a new release note for this PR.

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py (hunk 5 above)
Comment (adjs) [reply qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py:86]: I kept it public, but we can easily make its private and delete the release note of #9688.

**Comment by adjs** [comment]
> * Why don't the gates `mcrx`, `mcry` and `mcrz` appear in the circuit library?
>   https://qiskit.org/documentation/apidoc/circuit_library.html
>   Should they be added to the init file?
>   https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/circuit/library/__init__.py
> * Why do the `mcry` gate have a `q_ancillae` parameter and `mcrx` and `mcrz` do not have it?

I do not know the answers for these questions. I think this could be discussed in another PR, because these points are also present in the main branch.

**Comment by adjs** [comment]
> Thanks for this addition, overall the fix looks great! I left some comments below, mainly formalities.
> 
> If possible I think we would like to include this in the upcoming release, so let us know if there's anything we can help you with 🙂

I am not able to resolve the qpy_compat tests. `./run_experiments` doesn't work on my computer and I couldn't find documentation on these tests. Could you fix the compatibility tests?

**Comment by adjs** [comment]
@ShellyGarion and @Cryoris thanks for the review.

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py (hunk 1 above)
Comment (adjs) [reply qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py]: I added back the `use_basis_gates: bool = False`. Im not sure if this PR is the right place to include the deprecation note.

**Comment by mtreinish** [comment]
I looked at the qpy test failures locally (fwiw all you need is to have gnu parallel installed and run `./run_tests.sh` from the `test/qpy_compat` dir) the root cause of the failure is the changes to the generic `.control()` definition. This is causing a divergence from the nested definition of the custom controlled gate from the qpy files generate in the old release and the expected circuit we generate in the current main version of the code. I think if we're saying the definition prior to this PR was incorrect and needs to be different the only path forward here is to change https://github.com/Qiskit/qiskit-terra/blob/main/test/qpy_compat/test_qpy.py#L605-L610 to only run `generate_controlled_gates()` and `generate_open_controlled_gates()` if ` version_parts >= (0, 25, 0)` because we're not going to be able to validate the equivalence of the stored qpy prior to this change otherwise.

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py [hunk 10]
```diff
@@ -336,13 +342,16 @@ def mcrz(
...
-        _apply_mcu_graycode(
-            self, 0, 0, lam_step, control_qubits, target_qubit, use_basis_gates=use_basis_gates
-        )
+        mcsu2_real_diagonal(self, RZGate(lam).to_matrix(), control_qubits, target_qubit)
+
+    if use_basis_gates:
+        # pylint: disable=cyclic-import
+        from qiskit import transpile
+
+        self = transpile(self, basis_gates=["p", "u", "cx"], optimization_level=0)
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py:354]: I don't think this is quite correct, sorry if I didn't express it correctly before. With this code, the whole circuit will get transpiled to the specified basis -- but we only want the newly added mcrz to be in that basis 🙂 maybe we can just change `mcsu2_real_diagonal` to return the MCRZ subcircuit and then transpile that?

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py (hunk 5 above)
Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py:86]: I think it would be better if it's private, this way we can still move it e.g. to the synthesis module later, or turn it into a class without having to deprecate it. Since the release is imminent, I'll open a short separate PR for that 🙂

File: qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py (hunk 10 above)
Comment (adjs) [reply qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py]: Hi @Cryoris , I removed the transpile call. 
I think that we should not transpile the MCRZ circuit here, because of the transpile cost with a higher number of controls.

Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py]: I agree it's not the most efficient way, but we should ensure that `use_basis_gates=True` will still behave as it did before 🙂 If we want to avoid transpiling, would it be possible to express `mcsu2` in the `P U CX` basis?

**Comment by adjs** [comment]
Thanks @Cryoris. Your modifications LGTM.

**Review by mtreinish — APPROVED** [review]
LGTM too
